### PR TITLE
[fix] Never block threads if a non-IOException is thrown

### PR DESCRIPTION
**Pull request by j-baker**
Instead, just wrap in an IOException

We should not be catching errors, but they will just get blackholed.

## Before this PR
If a remote call throws an exception, the Conjure call will block forever. 

## After this PR
Now, fail the call with an exception.

## Possible downsides?
People relying on threads getting permanently blocked for control flow purposes will see a break.

### CODE REVIEW

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/CatchThrowableInterceptor.java
Comment (carterkozak) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/CatchThrowableInterceptor.java]: Can we please use SafeIOException here?

Comment (iamdanfox) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/CatchThrowableInterceptor.java]: Think you want a SafeIoException here to make sure the method doens't get redacted

File: okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/CatchThrowableInterceptorTest.java
Comment (carterkozak) [inline okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/CatchThrowableInterceptorTest.java:50]: May want to add something like this to verify that it does actually fail gracefully in practice:
https://github.com/palantir/conjure-java-runtime/pull/926/files

Comment (carterkozak) [reply okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/CatchThrowableInterceptorTest.java:50]: Eh that test is outdated, tracer has been refactored to catch+log since then. disregard.

**Review by iamdanfox — APPROVED** [review]

File: okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/OkHttpClientsTest.java
Comment (iamdanfox) [inline okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/OkHttpClientsTest.java:779]: @carterkozak I've jammed an IllegalStateException in here to make sure the interceptor is actually wire up - what do you think?

Comment (carterkozak) [reply okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/OkHttpClientsTest.java:779]: lgtm as long as this test fails without CatchThrowableInterceptor!

Comment (carterkozak) [reply okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/OkHttpClientsTest.java:779]: s/fails/hangs

**Comment by iamdanfox** [comment]
Seems like okhttp maintainers are aware of the issue, should be safe to remove this code once this issue is resolved: https://github.com/square/okhttp/issues/5151

File: okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/OkHttpClientsTest.java
Comment (iamdanfox) [reply okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/OkHttpClientsTest.java:779]: Yep, confirmed you get an exception from a dispatcher thread:

```
17:43:24.535 [remoting-okhttp-dispatcher-0] ERROR com.palantir.conjure.java.okhttp.OkHttpClients - An exception was uncaught in an execution thread. This likely left a thread blocked, and is as such a serious bug which requires debugging.
java.lang.IllegalStateException: I am not an IOException
	at com.palantir.conjure.java.okhttp.OkHttpClientsTest$5.record(OkHttpClientsTest.java:780)
	at com.palantir.conjure.java.okhttp.InstrumentedInterceptor.intercept(InstrumentedInterceptor.java:64)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.palantir.conjure.java.okhttp.ConcurrencyLimitingInterceptor.intercept(ConcurrencyLimitingInterceptor.java:74)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.palantir.conjure.java.okhttp.DispatcherTraceTerminatingInterceptor.lambda$intercept$0(DispatcherTraceTerminatingInterceptor.java:32)
	at com.palantir.tracing.AsyncTracer.withTrace(AsyncTracer.java:81)
	at com.palantir.conjure.java.okhttp.DispatcherTraceTerminatingInterceptor.intercept(DispatcherTraceTerminatingInterceptor.java:32)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:264)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:210)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at com.palantir.tracing.DeferredTracer.withTrace(DeferredTracer.java:108)
	at com.palantir.tracing.Tracers$TracingAwareCallable.call(Tracers.java:338)
	at com.palantir.tracing.WrappingExecutorService.lambda$wrapTask$0(WrappingExecutorService.java:67)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

and the test hangs.

**Comment by carterkozak** [comment]
lgtm, this fixes #925 👍

**Pull request merged**
